### PR TITLE
Install runner

### PR DESCRIPTION
**Pull request by BugDiver**
This PR address issue #335 along with some refactoring and enhancement

* Support for stand-alone `.spec` files (only ones which are related to valid gauge project)
* Propagate error to the user if  debugging cannot start

Technical changes
* Introduce `CLI` class which holds information about `gauge` and `maven`
* CLI also perform tasks related to gauge and maven command line like `install`, version etc. (Still some things are outside which can be moved).
* Introduce `GaugeProject` which holds information about the gauge project like `runner`, plugins, which command to use for execution,  is a maven project etc.
* GaugeProject can tell if a file belongs to it.
* Introduce `GaugeClients` to reduce the computation of gauge project again and again. Example: If a run is initiated for a spec, the project root already computed and cached so no need to compute again. It holds a `{project: GaugeProject, client: LanguageClient}` type against the project root.
* Removed `GaugeVersionInfo` in favour of `CLI`
* Removed util methods in favour of `CLI` and `GaugeProject` class methods.

### CODE REVIEW

File: src/cli.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        let cm = this._gaugeCommitHash && `Commit Hash: ${this._gaugeCommitHash}` || '';
+        let plugins = this._gaugePlugins
+            .map((p: any) => p.name + ' (' + p.version + ')')
+            .join('\n');
+        plugins = `Plugins\n-------\n${plugins}`;
+        return `${v}\n${cm}\n\n${plugins}`;
+    }
+}
+
+export function getCLI() {
```
Comment (sriv) [inline src/cli.ts]: call this `Instance`? ex. `CLI.Instance()`?

File: src/constants.ts [hunk 2]
```diff
@@ -72,6 +72,11 @@ export enum GaugeRequests {
     SpecDirs = 'gauge/specDirs'
 }
 
+export enum GaugeRunners {
+    Java = "java",
+    Dotnet = "Dotnet"
```
Comment (sriv) [inline src/constants.ts]: Why is `Dotnet` in caps?

File: src/execution/gaugeExecutor.ts [hunk 3]
```diff
@@ -41,53 +45,64 @@ export class GaugeExecutor extends Disposable {
...
-                        }
-                        if (env.DEBUGGING && lineText.indexOf(NO_DEBUGGER_ATTACHED) >= 0) {
-                            window.showErrorMessage("No debugger attached. Stopping the execution");
-                            this.cancel();
-                        }
+                    this.childProcess.stdout.on('data', this.filterStdoutDataDumpsToTextLines((lineText: string) => {
+                        chan.appendOutBuf(lineText);
+                        let lineTexts = lineText.split("\n");
+                        lineTexts.forEach((lineText) => {
+                            if (lineText.indexOf(REPORT_PATH_PREFIX) >= 0) {
```
Comment (sriv) [inline src/execution/gaugeExecutor.ts]: these `if` could be broken out to a few `LineTextProcessors` ?

File: src/gaugeProject.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            this._plugins = manifest.Plugins;
+        }
+    }
+
+    public getExecutionCommand(cli: CLI): string {
+        if (this.isMavenProject()) return cli.mavenCommand();
+        return cli.gaugeCommand();
+    }
+
+    public isMavenProject() {
```
Comment (sriv) [inline src/gaugeProject.ts]: perhaps `MavenProject` can inherit from`GaugeProject` and a factory to return the right project type?

File: src/gaugeWorkspace.ts [hunk 5]
```diff
@@ -151,18 +158,34 @@ export class GaugeWorkspace extends Disposable {
...
         languageClient.start();
-        languageClient.onReady().then(() => { this.setLanguageId(languageClient, folderPath); });
+        return languageClient.onReady().then(() => { this.setLanguageId(languageClient, project.root()); });
+    }
+
+    private async installRunnerFor(project: GaugeProject): Promise<any> {
+        const language = project.language();
+        if (this.cli.isPluginInstalled(language)) return;
+        let message = `The project ${path.basename(project.root())} requires gauge ${language} to be installed. ` +
+            "Do you like to install?";
```
Comment (sriv) [inline src/gaugeWorkspace.ts]: "Do you like to install?" -> "Would you like to install it?"

File: src/execution/debug.ts [hunk 6]
```diff
@@ -80,13 +85,16 @@ export class GaugeDebugger {
...
         let env = Object.create(process.env);
         if (this.debug) {
             env.DEBUGGING = true;
+            env.use_nested_specs = "false";
+            env.SHOULD_BUILD_PROJECT = "true";
             return getPort({ port: DEBUG_PORT }).then((port) => {
-                if (isProjectLanguage(projectRoot, 'dotnet')) env.GAUGE_CSHARP_PROJECT_CONFIG = "Debug";
-                if (isProjectLanguage(projectRoot, 'java')) env.GAUGE_DEBUG_OPTS = port;
+                if (this.config.getProject().isProjectLanguage(GaugeRunners.Dotnet))
+                    env.GAUGE_CSHARP_PROJECT_CONFIG = "Debug";
```
Comment (sriv) [inline src/execution/debug.ts]: use braces for `if`

**Comment by zabil** [comment]
Approving as it was already approved by @sriv

**Pull request merged**